Objects/dictobject.c

```diff
@@ -234,6 +235,14 @@ set_values(PyDictObject *mp, PyDictValues *values)
 
 #endif
 
+#define STORE_KEY(ep, key) FT_ATOMIC_STORE_PTR_RELAXED(ep->me_key, key)
+#define STORE_VALUE(ep, value) FT_ATOMIC_STORE_PTR_RELAXED(ep->me_value, value)
+#define STORE_SPLIT_VALUE(mp, idx, value) FT_ATOMIC_STORE_PTR_RELAXED(mp->ma_values->values[idx], value)
```

Comment:
I think we want "release", although I'm not 100% sure. It's possible that the synchronization from the incref is sufficient even with relaxes stores

Reply:
Maybe I've failed to capture exactly what you're concerned about, but I think the synchronization around keys and values provides what we need: https://gist.github.com/DinoV/fd83e80e5eb37ffa197dadd9791cca6e because reads to ma_keys and ma_values are fully sequential, and writes to them are done with release stores.

Reply:
And I think we'll be forced to lock on both threads because `_Py_TryXGetRef` is going to fail on the value that's not yet shared across threads and we'll be forced into the locking code path.

Reply:
After thinking about this a bit more, I think (3) really needs to be at least "release".

First, contrary to what I wrote above, the mutex locking isn't sufficient because a seq-cst operation on a variable is weaker than a seq-cst fence -- it doesn't 

I don't think we can rely on the `_Py_TryXGetRef` failing either. The object might have a non-zero shared refcount for a number of reasons.

Here's a simplified model. If you change the relevant "memory_order_release" to "memory_order_relaxed" it will fail:
https://github.com/colesbury/c11-model-checker/blob/cpython-models/test/dict_value.c